src/BizHawk.Client.Common/Api/WebSocketServer.cs

```diff
@@ -1,14 +1,215 @@
+		{
+			if (_running)
+			{
+				throw new InvalidOperationException("Server has already been started");
+			}
+			_cancellationToken = cancellationToken;
+			_running = true;
+
+			clientRegistrationListener.Start();
+			await ListenForAndRegisterClients();
+		}
+
+		private async Task ListenForAndRegisterClients()
+		{
+			while (_running && !_cancellationToken.IsCancellationRequested)
+			{
+				var context = await clientRegistrationListener.GetContextAsync();
+				if (context is null) return;
+
+				if (!context.Request.IsWebSocketRequest)
+				{
+					context.Response.Abort();
+					return;
+				}
+
+				var webSocketContext = await context.AcceptWebSocketAsync(subProtocol: null);
+				if (webSocketContext is null) return;
+				RegisterClient(webSocketContext.WebSocket);
+			}
+		}
+
+		private void RegisterClient(WebSocket newClient)
+		{
+			string clientId = Guid.NewGuid().ToString();
+			clients.Add(clientId, newClient);
+			_ = Task.Run(() => ClientMessageReceiveLoop(clientId), _cancellationToken);
+		}
+
+		private async Task ClientMessageReceiveLoop(string clientId)
+		{
+			byte[] buffer = new byte[1024];
+			var messageStringBuilder = new StringBuilder(2048);
+			var client = clients[clientId];
+			while (client.State == WebSocketState.Open && !_cancellationToken.IsCancellationRequested)
+			{
+				ArraySegment<byte> messageBuffer = new(buffer);
+				var receiveResult = await client.ReceiveAsync(messageBuffer, _cancellationToken);
+				if (receiveResult.Count == 0)
+					return;
+
+				messageStringBuilder.Append(Encoding.ASCII.GetString(buffer, 0, receiveResult.Count));
+				if (receiveResult.EndOfMessage)
+				{
+					string messageString = messageStringBuilder.ToString();
+					messageStringBuilder = new StringBuilder(2048);
+
+					try
+					{
+						var request = JsonSerde.Deserialize<RequestMessageWrapper>(messageString);
+						await HandleRequest(clientId, request);
+					}
+					catch (Exception e)
+					{
+						// TODO proper logging
+						Console.WriteLine("Error deserializing message {0}", e);
+						await SendClientGenericError(clientId);
+					}
+				}
+			}
+		}
+
+		private async Task HandleRequest(string clientId, RequestMessageWrapper request)
+		{
+			try
+			{
+				switch (request.Topic)
+				{
+					case Topic.Error:
+						// clients arent allowed to publish to this topic
+						await SendClientGenericError(clientId);
+						break;
+
+					case Topic.Registration:
+						await HandleRegistrationRequest(clientId, request.Registration!.Value);
+						break;
+
+					case Topic.Echo:
+						await HandleEchoRequest(clientId, request.Echo!.Value);
+						break;
+				}
+
+			}
+			catch (Exception e)
+			{
+				// this could happen if, for instance, the client sent a registration request to the echo topic, such
+				// that we tried to access the wrong field of the wrapper
+				// TODO proper logging
+				Console.WriteLine("Error handling message {0}", e);
```

Comment:
Would love a suggestion on how to properly do logging (both debug and otherwise)